actionpack/lib/action_dispatch/http/url.rb

```diff
@@ -11,6 +11,24 @@ module URL
 
       mattr_accessor :secure_protocol, default: false
       mattr_accessor :tld_length, default: 1
+      mattr_accessor :domain_extractor
+
+      class DomainExtractor # :nodoc:
+        attr_reader :host, :tld_length
+
+        def initialize(host, tld_length)
```

Comment:
I'm not sure we need an instance. Can we pass host and tld_length to domain and subdomains?

Reply:
I think it can be a good thing to have an instance and treat it like a data structure. It could handle things like caching and other complex logic plus reusability as well.

Reply:
YAGNI. We really don't need to think about caching and reusability. If your implementation needs an instance, you can pass an instance to it, but by default we only expect an object that respond to `domain(host, tld_length)` and `subdomains(host, tld_length)`.

Reply:
I don't think we should create unnecessary instances by default in the framework, we are generating unnecessary allocation.

Reply:
@rafaelfranca Got it. Take a look at the recent changes plz.